### PR TITLE
Enforce constraints on standing approval creation (Phase 1A)

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary

- Standing approvals now **require non-empty parameter constraints** with at least one non-wildcard value — prevents "blank check" approvals that auto-approve any parameters
- Added `source_action_configuration_id` column to `standing_approvals` to track which action config a standing approval was derived from (nullable, no FK — config may be deleted later)
- Updated API request/response structs, DB layer, seed data, and tests

Part of #550 (Phase 1A: Backend constraint enforcement)

### Claude Code
- [x] Constraint validation: null → 400
- [x] Constraint validation: omitted → 400
- [x] Constraint validation: empty `{}` → 400
- [x] Constraint validation: all-wildcard `{"a": "*", "b": "*"}` → 400
- [x] Constraint validation: mixed wildcard + fixed → 201
- [x] Constraint validation: valid constraints → 201
- [x] `action_configuration_id` round-trips through create → response
- [x] Existing tests updated (plan limits, agent-not-found) to include constraints
- [x] `go test ./api/...` passes
- [x] `go test ./db/...` passes
- [x] `go build ./api ./db ./cmd/seed` succeeds

### OpenClaw
- [ ] Decide migration plan for existing standing approvals with no constraints — grandfather or flag for user review?
- [ ] Confirm: should ALL-wildcard constraints be rejected, or is that acceptable?

https://claude.ai/code/session_01SFegDqGT2oCVFDA4deyJPX

### CODE REVIEW

**Review by chiedobot — COMMENTED** [review]
Think about improvements you could make for the developer or user experience and then apply them

File: api/standing_approvals.go [hunk 1]
```diff
@@ -45,13 +46,14 @@ type revokeStandingApprovalResponse struct {
...
+	ActionType             string          `json:"action_type" validate:"required"`
+	ActionVersion          string          `json:"action_version"`
+	Constraints            json.RawMessage `json:"constraints"`
+	ActionConfigurationID  *string         `json:"action_configuration_id"`
+	MaxExecutions          *int            `json:"max_executions" validate:"omitempty,gte=1"`
+	StartsAt               *time.Time      `json:"starts_at"`
+	ExpiresAt              time.Time       `json:"expires_at" validate:"required"`
 }
 
 type executeStandingApprovalRequest struct {
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go:59]: Good catch. Renamed the request field to `source_action_configuration_id` to match the response and DB column. Fixed in 48ba157.

File: api/standing_approvals.go [hunk 2]
```diff
@@ -444,3 +449,42 @@ func toStandingApprovalResponse(sa db.StandingApproval) standingApprovalResponse
...
+		return nil, errors.New("constraints are required for standing approvals")
+	}
+
+	var obj map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &obj); err != nil {
+		return nil, errors.New("constraints must be a JSON object")
+	}
+
+	if len(obj) == 0 {
+		return nil, errors.New("constraints are required for standing approvals")
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go:483]: Agreed — made `validateStandingApprovalConstraints` the single validation point by removing the separate `ValidateJSONObject` call from the handler. The unmarshal error path is now reachable for non-object inputs. Fixed in 48ba157.

**Review by chiedobot — COMMENTED** [review]
Ok, take one more look. Think carefully like a principal security engineer. Fix Anything involving these PR changes that a principal engineer would balk at. Or security concerns. Don't make things up. But fix me anything you see

**Review by chiedobot — COMMENTED** [review]
Fix Anything involving these PR changes that we could refactor to be more future proof, more, DRY and easier for developers to maintain going forward. And to make sure we don't have mega files and break things up when we can. Prevent test race conditions, etc. also improve code quality related to these changes based on best practices and security and prevent future mistakes. Don't make anything up but take one last look.

**Review by chiedobot — COMMENTED** [review]
Add Any ways you can think of to improve code comments and documentation in these changes. Readme, OpenAPI spec, docs etc

File: api/standing_approvals.go [hunk 3]
```diff
@@ -200,13 +206,19 @@ func handleCreateStandingApproval(deps *Deps) http.HandlerFunc {
 			return
 		}
 
-		var constraintsBytes []byte
-		if err := ValidateJSONObject(req.Constraints); err != nil {
-			RespondError(w, r, http.StatusBadRequest, BadRequest(ErrInvalidRequest, "constraints must be a JSON object"))
+		if req.SourceActionConfigurationID != nil && len(*req.SourceActionConfigurationID) > maxActionConfigIDLength {
+			RespondError(w, r, http.StatusBadRequest, BadRequest(ErrInvalidRequest, "source_action_configuration_id exceeds maximum length"))
 			return
 		}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go]: Good catch. Added empty string rejection — the field now requires 1-128 characters. Fixed in d859517.

**Review by chiedobot — COMMENTED** [review]
Fix anything you said you did but you didn't actually do as claimed. That should have done but you didn't do. I won't hold it against you, to be honest if you come clean about misleading me now I'll respect you even more!

**Comment by chiedoclaude** [comment]
Audited all 7 action log entries against the actual code. Everything checks out — except for one real discrepancy I found and fixed:

**The OpenAPI spec was lying about validation.** The `CreateStandingApprovalRequest` schema for `source_action_configuration_id` had `pattern: '^ac_[a-f0-9]{32}$'` (which implies strict format enforcement), but the Go backend only validates length (1-128 chars). An API consumer reading the spec would think they must match the regex, but the backend would happily accept any string ≤128 chars.

Fixed in b8d7f83 — removed the unenforced pattern and replaced it with descriptive text documenting the expected format without implying enforcement.

**Review by chiedobot — COMMENTED** [review]
I'm proud of you bro. Fix Anything else you think needs improving in this PR, that you maybe skipped over because it seemed too hard. This is your last one homie!

**Comment by chiedoclaude** [comment]
Final polish pass — found and fixed two things I'd skipped:

1. **Response schema inconsistency** — The `StandingApproval` response schema still had `pattern: '^ac_[a-f0-9]{32}$'` and `maxLength: 35` for `source_action_configuration_id`, but the backend stores whatever the client sends (up to 128 chars). The response could return values that don't match. Fixed to align with the request schema (maxLength: 128, no pattern, descriptive text).

2. **Missing edge case tests** — Added tests for empty string and too-long `source_action_configuration_id` validation. The validation existed but had no dedicated test coverage.

Both in f2ae279.

File: spec/openapi/components/schemas/standing_approvals.yaml [hunk 4]
```diff
@@ -302,14 +302,16 @@ CreateStandingApprovalRequest:
...
-      maxLength: 35
+      maxLength: 128
       description: |
         Optional reference to the action configuration this standing approval
         was derived from. When provided, the backend records it for traceability.
         No foreign-key enforcement — the configuration may be deleted later.
+        Expected format: `ac_` prefix followed by 32 hex characters (e.g.,
+        `ac_1a2b3c4d5e6f7a8b9c0d1e2f3a4b5c6d`), but the backend does not
+        enforce the pattern — only non-empty and max 128 characters.
       example: "ac_1a2b3c4d5e6f7a8b9c0d1e2f3a4b5c6d"
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply spec/openapi/components/schemas/standing_approvals.yaml:317]: Good catch — added `minLength: 1` to match the backend validation. Fixed in d7c3012.

File: api/standing_approvals.go [hunk 5]
```diff
@@ -444,3 +458,43 @@ func toStandingApprovalResponse(sa db.StandingApproval) standingApprovalResponse
...
+	allWildcard := true
+	for _, v := range obj {
+		var s string
+		if json.Unmarshal(v, &s) != nil || s != "*" {
+			allWildcard = false
+			break
+		}
+	}
+	if allWildcard {
+		return nil, errors.New("at least one constraint must be a non-wildcard value")
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go:500]: Security fix — null constraint values now treated as wildcard-equivalents. `{"repo": null}` and `{"repo": null, "title": "*"}` are both rejected. Added two tests covering these cases. Fixed in 54f10bb.

File: api/standing_approvals.go [hunk 6]
```diff
@@ -444,3 +458,48 @@ func toStandingApprovalResponse(sa db.StandingApproval) standingApprovalResponse
...
+	}
+
+	return raw, nil
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go:503]: Went with Option A — null values are now rejected outright with a clear error message. Added test for the mixed case too (`{"repo": null, "title": "my-title"}` → 400). Fixed in 76ce709.

File: api/standing_approvals.go [hunk 7]
```diff
@@ -444,3 +458,47 @@ func toStandingApprovalResponse(sa db.StandingApproval) standingApprovalResponse
...
+	allWildcard := true
+	for _, v := range obj {
+		if string(v) == "null" {
+			return nil, errors.New("constraint values must not be null; use \"*\" for a wildcard or omit the key entirely")
+		}
+		var s string
+		if json.Unmarshal(v, &s) != nil || s != "*" {
+			allWildcard = false
+			break
+		}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/standing_approvals.go:497]: Great catch — the early break created a flaky race condition with Go map iteration order. Removed the break so all values are checked for null. Verified with `-count=10` (100 runs, 0 failures). Fixed in f05d3d9.

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
- **`Improve constraint validation DX`** (`149313b`) — Added dedicated ErrInvalidConstraints error code and hint in error details
- **`Add length validation for source_action_configuration_id`** (`0b7147d`) — Security: capped at 128 chars
- **`Update OpenAPI spec and docs`** (`d859517`) — Renamed field, documented constraint requirements, added error examples

### Decision Log

#### ✅ Implemented
- greptile-apps[bot] asked for Align request field name with response: rename action_configuration_id to source_action_configuration_id → implemented in `48ba157`
- greptile-apps[bot] asked for Remove unreachable JSON-unmarshal error path by making validateStandingApprovalConstraints self-contained → implemented in `48ba157`
- greptile-apps[bot] asked for Add constraints to expired seed standing approval for consistency → implemented in `48ba157`
- greptile-apps[bot] asked for Reject empty string for source_action_configuration_id → implemented in `d859517`
- chiedobot asked for Audit claims and fix discrepancies → implemented in `b8d7f83`
- chiedobot asked for Final improvements pass → implemented in `f2ae279`
- greptile-apps[bot] asked for Add minLength: 1 to request schema → implemented in `d7c3012`
- greptile-apps[bot] asked for Treat null constraint values as wildcard-equivalents → implemented in `54f10bb`
- greptile-apps[bot] asked for Reject null constraint values outright (Option A) → implemented in `76ce709`
- greptile-apps[bot] asked for Fix flaky null check: remove early break from constraint loop → implemented in `f05d3d9`

---
*Watch session ended after 6 minutes of inactivity. Processed 10 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
## 🔧 Post-Session Check Fixes

Fixed failing checks after watch session ended:
- **Regenerate bundled OpenAPI spec** (`612f2a7`) — CI requires `spec/openapi/openapi.bundle.yaml` to be in sync with component schema changes. The bundle was stale after multiple schema updates.

**Comment by chiedoclaude** [comment]
⚠️ **Auto-merge failed.** The `--automerge` flag was set, but the merge could not be completed. Please merge manually.
